packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/utils/ipData.ts

```diff
@@ -0,0 +1,68 @@
+import { Logger } from "pino"
+import { Request } from "express"
+import axios from "axios"
+import { config } from "../config"
+import { logger } from "../logger"
+
+export type LocationData = { city: string, region: string, country: string } | {}
+
+// gets ip data from api.ipdata.co, returns an empty object {} if api key not configured or an error occurs
+export const getIpData = async (logger: Logger, ip: string): Promise<LocationData> => {
+    const ipdataApiKey = config.ipdataApiKey
+    if (ipdataApiKey === null) {
+        logger.warn({}, "ip data requested but api key not configured")
+        return {}
+    }
+    const url = `https://api.ipdata.co/${ip}?api-key=${config.ipdataApiKey}`
+    try {
+        const response = await axios.get(url)
+            .then(({ data: { city, region, country_name } }: { data: { city: string, region: string, country_name: string } }) => {
+                return { city, region, country: country_name }
+            })
+        return response
+    } catch (e: unknown) {
+        logger.error({ error: e }, "error requesting ip data")
+        return {}
+    }
+}
+
+// gets ip data from api.ipdata.co, returns an empty object {} if api key not configured or an error occurs
+export const getRequestIpData = async (logger: Logger, req: Request): Promise<LocationData> => {
+    try {
+        const ip = getIP(req)
+        return getIpData(logger, ip)
+    } catch (e) {
+        logger.error({ e }, "error requesting ip data")
+        return {}
+    }
+}
+
+// Utility to gather the IP from an incoming express request.
+// Prioritizes 'X-Forwarded-For' if it exists. Otherwise returns the request objects ip.
+// Throws if neither can be found.
+export const getIP = (req: Request): string => {
+    const ip = req.ip
+    const forwardedFor = req.get('X-Forwarded-For')
```

Comment:
we should update this logic because X-Forwarded-For may actually be an array (gross I know).

can we roll with a library instead? https://github.com/pbojinov/request-ip it seems pretty reasonably good

```
const requestIp = require('request-ip');
app.use(requestIp.mw())

app.use(function(req, res) {
    const ip = req.clientIp;
    res.end(ip);
});
```
in the `index.ts` file

Reply:
sure!